### DISCUSSION TITLE
Why Sqlalchemy and Alembic don't detect my Tables without import them in the main file or env.py(For Alembic) ?

### DISCUSSION THREAD

**Discussion by gikpro** [sqlalchemy/sqlalchemy#7541] [Usage Questions]
I am working on a FASTAPI project, i am using Sqlalchemy 1.4.28 and Alembic 1.7.5.
My project packages are like that
----| Project
------| Models ( Contains ORM class of my tables that depend on the declarative_base instance named "Base")
--------| tab1.py
--------| tab2.py
-------| database.py ( Contains all initialisation of the connection with the database using sqlalchemy)
-------| .....
-------| run.py ( i import engine, Base variables from database.py, and write Base.metadata.create_all (engine)

When i start the program, my sqlite database is created but it is empty.
I had the same problem with the autogeneration of my Tables using Alembic despite I followed the documentation of Alembic. I have added "target_metadata = Base.metadata" in the env.py.
I spent days looking for a solution to this problem, until I had the idea to import the Tables into my run.py file (when I am not using Alembic) and import the tables in the env.py file (when I use autogenerate from Alembic).
And as if by magic everything works.

I would like to know what happened, is it a problem with the structure of my project that is preventing Alembic and SqlAlchemy from detecting my Tables? And for my solution to import my Tables without using them, is that a bad solution? Thank you

**Comment by CaselIT** [comment]
Hi,

This is just how python import works. A module is imported and executed only if explicitly imported and a SQLAlchemy table / model is registered only if it's module is executed.
You can find more information on how it python import work here https://docs.python.org/3/reference/import.html

**Reply by gikpro** [reply]
Hi,
I know how import work in python. I asked this question because in my Flask's projects i never do that. Flask-migrate, Flask-SqlAlchemy may be have do some tricks in background.

**Reply by CaselIT** [reply]
I've never used them so I don't know how they work.
SQLAlchemy does not automatically import modules, they must be manually imported.

**Comment by cesarrodrig** [comment]
I also had the same problem. I was setting `target_metadata = Base.metadata` while having different tables defined in other files (while inheriting from `Base`). It was until I imported the tables manually that they were detected by `alembic`. This should be added as part of the `Auto Generating Migrations` docs.
